app/src/main/java/com/alphawallet/app/entity/tokens/Attestation.java

```diff
@@ -0,0 +1,131 @@
+package com.alphawallet.app.entity.tokens;
+
+import static com.alphawallet.app.repository.TokensRealmSource.attestationDatabaseKey;
+import static com.alphawallet.app.repository.TokensRealmSource.databaseKey;
+
+import android.text.TextUtils;
+
+import com.alphawallet.app.entity.ContractType;
+import com.alphawallet.app.repository.entity.RealmAttestation;
+import com.alphawallet.token.entity.AttestationValidation;
+import com.alphawallet.token.entity.AttestationValidationStatus;
+import com.alphawallet.token.entity.TokenScriptResult;
+import com.alphawallet.token.tools.Numeric;
+
+import org.web3j.abi.datatypes.Type;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.Map;
+
+/**
+ * Created by JB on 23/02/2023.
+ */
+public class Attestation extends Token
+{
+    private final byte[] attestation;
+    private BigInteger attestationId;
+    private String attestationSubject;
+    private String issuerKey;
+    private String issuerAddress;
+    private long validFrom;
+    private long validUntil;
+    private Map<String, Type<?>> additionalMembers;
+    private boolean isValid;
+    private ContractType baseTokenType = ContractType.ERC721; // default to ERC721
+
+    //TODO: Supplemental data
+
+    public Attestation(TokenInfo tokenInfo, String networkName, byte[] attestation)
+    {
+        super(tokenInfo, BigDecimal.ZERO, System.currentTimeMillis(), networkName, ContractType.ATTESTATION);
+        this.attestation = attestation;
+        setAttributeResult(BigInteger.ZERO, new TokenScriptResult.Attribute("attestation", "attestation", BigInteger.ZERO, Numeric.toHexString(attestation)));
+    }
+
+    public byte[] getAttestation()
+    {
+        return attestation;
+    }
+
+    public void handleValidation(AttestationValidation attValidation)
+    {
+        if (attValidation == null)
+        {
+            return;
+        }
+
+        attestationId = attValidation._attestationId;
+        attestationSubject = attValidation._subjectAddress;
+        issuerAddress = attValidation._issuerAddress;
+        isValid = attValidation._isValid;
+        additionalMembers = attValidation.additionalMembers;
+        issuerKey = attValidation._issuerKey;
+    }
+
+    public AttestationValidationStatus isValid()
+    {
+        if (!isValid)
+        {
+            return AttestationValidationStatus.Expired;
+        }
+        else if (!TextUtils.isEmpty(attestationSubject) && !attestationSubject.equalsIgnoreCase(getWallet()))
```

Comment:
These `else` are redundant.

Reply:
Hi @seabornlee could you elaborate? This function returns a validation status which could either be "Pass" if everything checks out, or it could have failed in one of 3 ways:

- The attestation has expired (outside of the time gate).
- The attestation subject (who it was issued to) does not match the importing wallet.
- The issuer of the attestation is not correct (it was signed by an unrecognised key).